keps/sig-apps/20190318-PodDisruptionBudget-graduation-to-stable.md

```diff
@@ -0,0 +1,133 @@
+---
+title: graduate-pod-disruption-budget-to-stable
+authors:
+  - "@bsalamat"
+owning-sig: sig-apps
+participating-sigs:
+  - sig-scheduling
+reviewers:
+  - "@liggitt"
+  - "@kow3ns"
+approvers:
+  - "@kow3ns"
+  - "@liggitt"
+editor: TBD
+creation-date: 2019-03-18
+last-updated: 2019-03-18
+status: implementable
+see-also:
+  - 
+replaces:
+superseded-by:
+---
+
+# Graduate PodDisruptionBudget to stable
+
+## Table of Contents
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+
+## Summary
+
+[Pod Disruption Budget (PDB)](https://kubernetes.io/docs/tasks/run-application/configure-pdb/)
```

Comment:
@bsalamat @liggitt - I've been thinking a lot about this topic when working on ClusterAPI.  

Because the object is very loosely coupled with Pods, and only via label selector, is there any reason why we couldn't have it be more generic and remove the the 'Pod' from the resource kind?  The link via label denotes the resource that it is tied to.  

As a concrete example, we would like to tie disruption budgets to other resources for cluster management. 
/cc @ncdc @detiber

Reply:
for the other scenarios, is delete the standard definition of disruption?

Reply:
Agree with disruption budgets to machines.  There's two orthogonal dimensions and use cases.

Reply:
Would just using label selector cause confusion here? Would there be a case where we would want to match both Machines and Pods with the same selector? Or would it be on the consumer to differentiate the type to associate with the label selector?

As to the specific use case around Machines, would it make sense to have the disruption budget exist for the Machines or the underlying Nodes instead? If what we are striving for is to ensure a certain level of capacity is available during an upgrade, then it seems to me that we are more concerned with the Nodes rather than the Machines themselves.

Reply:
we have protected a pool of machines by running a pod per node with anti-affinity rules setup, and then you can prevent disruption to more than N number of nodes using that pod as a guard.  its a bit of a hack, but it works.  in general, i would prefer we have pod disruption budget, and machine disruption budgets as separate objects, the use cases may vary slightly in practice.  similar to having machine sets and replica sets.

Reply:
+1 to having different disruption budget for pods and nodes for a few reasons:
1. PDB is enforced through `evict` subresource of Pods which is obviously not applicable to nodes.
1. If we used the same API for both objects, we would need to add a field to the API to specify the type of object that it applies to (node vs pod).
1. Implementation of Node Disruption Budget shares little with that of PDB. So, we won't save much on the coding side by combining the two APIs either.

Reply:
> I would prefer we have pod disruption budget, and machine disruption budgets as separate objects, the use cases may vary slightly in practice. similar to having machine sets and replica sets

I'd tend to agree. The concepts are similar, but the application, audience, and scope are different.

PDB's are namespaced, and write access can be granted to users with write access in a namespace.

Machine or Node disruption budgets would be cluster-scoped, and should only be writeable by the cluster admin.

Reply:
> Machine or Node disruption budgets would be cluster-scoped, and should only be writeable by the cluster admin.

That's not true, they would be namespace scoped in a mgmt cluster. 

FWIW I'm ok if we decide to not rationalize now, but I do think this is an area in the api where there is overlapping concepts and have (N) of the same type is also not good.

Reply:
> FWIW I'm ok if we decide to not rationalize now,

If we think we may change the API at some point, discussing it now is probably better than after the API is GA'ed.

Reply:
We have work-arounds on our side.